csm/core/agent/api.py

```diff
@@ -317,11 +320,47 @@ async def set_secure_headers(request, handler):
         SecureHeaders(csp=True).aiohttp(resp)
         return resp
 
+    # The following workaround is required to ensure the ElasticSearch mappings for
+    # `CsmAuditLogModel` will be in place before any audit log entries are generated.  The query
+    # will force the correct mappings to be created on ElasticSearch according to the corresponding
+    # CSM model if they are not yet in place.  This is required because new audit log entries are
+    # added to ElasticSearch by rsyslog and, in case mappings are not present, rsyslog will create
+    # new ones itself.  The newly created mappings will differ from the ones expected by our
+    # business logic, and we will fail to read data from ElasticSearch.  The operation should be
+    # executed only once for performance purposes.
+
+    # Simple flag to keep track of its initialization amidst the sea of static methods in this
+    # module.
+
+    class AuditLogDbInitialized:
+        status = False
+
+    # This function should be invoked before `Log.audit()`.  The call will be placed inside
+    # `rest_middleware()` because it's the closest async parent call to `Log.audit()`.  We load the
+    # database schema from the YAML file again to avoid having to inject the database provider
+    # through the class constructor.  It is a workaround after all.
+
+    @staticmethod
+    async def _initialize_audit_log_db():
+        if CsmRestApi.AuditLogDbInitialized.status:
+            return
+        db_config = Yaml(const.DATABASE_CONF).load()
+        db_config['databases']["es_db"]["config"][const.PORT] = int(
+            db_config['databases']["es_db"]["config"][const.PORT])
+        db_config['databases']["es_db"]["config"]["replication"] = int(
+            db_config['databases']["es_db"]["config"]["replication"])
+        conf = GeneralConfig(db_config)
+        db = DataBaseProvider(conf)
+        await db(CsmAuditLogModel).get(Query().limit(0))
+        CsmRestApi.AuditLogDbInitialized.status = True
+
+
     @staticmethod
     @web.middleware
     async def rest_middleware(request, handler):
         try:
             request_id = int(time.time())
+            await CsmRestApi._initialize_audit_log_db()
```

Comment:
Please have in mind that this will be called for every single API request, which is objectively worse than being called during startup as the original patch proposed.  It will also have a bigger impact on the first request received every time `csm_agent` comes back up.

@kupranay @pawankumarsrivastava, FYI.

Reply:
we should have a flag which will be false for the first time and get updated to true once db is initialized. 
"await CsmRestApi._initialize_audit_log_db()" should be called only once when csm-agent will get started and flag should be reset

Reply:
This change is already part of the patch.  Please check https://github.com/Seagate/cortx-manager/pull/462/files#diff-976c77a4ccd126b2da94f02da86d6fbd963c46d3c34ae2f82b9eb52d607a9981R332